File: backend-java/src/main/java/com/peopleminer/events/EventsGateway.java
package com.peopleminer.events;

import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Data;
import lombok.NoArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.messaging.simp.SimpMessagingTemplate;
import org.springframework.stereotype.Component;

@Component
@Slf4j
public class EventsGateway {

    private final SimpMessagingTemplate messagingTemplate;

    public EventsGateway(SimpMessagingTemplate messagingTemplate) {
        this.messagingTemplate = messagingTemplate;
    }

    public void sendProgress(JobProgress progress) {
        messagingTemplate.convertAndSend("/topic/crawl-progress", progress);
    }

    public void sendCandidateUpdate(CandidateUpdate update) {
        messagingTemplate.convertAndSend("/topic/candidate-update", update);
    }

    @Data
    @NoArgsConstructor
    @AllArgsConstructor
    @Builder
    public static class JobProgress {
        private String jobId;
        private String source;
        private String status; // "processing", "completed", "error", "finished", "scored", "scoring"
        private String message;
        private Integer found;
        private Integer newCount;
        private String error;
        private String candidateId;
        private Double score;
        private Integer scored;
        private Integer failed;
        private Double progress;
    }

    @Data
    @NoArgsConstructor
    @AllArgsConstructor
    @Builder
    public static class CandidateUpdate {
        private String id;
        private String githubUsername;
    }
}

File: backend/src/events/events.gateway.ts
import {
  WebSocketGateway,
  WebSocketServer,
  OnGatewayConnection,
  OnGatewayDisconnect,
} from '@nestjs/websockets';
import { Server, Socket } from 'socket.io';
import { Logger } from '@nestjs/common';

export interface JobProgress {
  jobId: string;
  source?: string;
  status: 'processing' | 'completed' | 'error' | 'finished' | 'scored' | 'scoring';
  message: string;
  found?: number;
  new?: number;
  error?: string;
  candidateId?: string;
  score?: number;
  scored?: number;
  failed?: number;
  progress?: number;
}

export type CrawlProgress = JobProgress;

@WebSocketGateway({
  cors: {
    origin: ['http://localhost:5173', 'http://localhost:3000'],
    credentials: true,
  },
})
export class EventsGateway implements OnGatewayConnection, OnGatewayDisconnect {
  @WebSocketServer()
  server: Server;

  private readonly logger = new Logger(EventsGateway.name);

  handleConnection(client: Socket) {
    this.logger.log(`Client connected: ${client.id}`);
  }

  handleDisconnect(client: Socket) {
    this.logger.log(`Client disconnected: ${client.id}`);
  }

  sendProgress(progress: JobProgress) {
    this.server.emit('crawl-progress', progress);
  }

  sendCandidateUpdate(candidate: { id: string; githubUsername: string }) {
    this.server.emit('candidate-update', candidate);
  }
}
